File: day16.c
#include <assert.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#define COUNT(a) (sizeof(a) / sizeof((a)[0]))

#define WS 8
#define HS 8
#define W (1 << WS)
#define H (1 << HS)
#define A(x, y) a[y][x]
char a[H][W];
int n;

const int dx[] = { 1, 0,-1, 0 };
const int dy[] = { 0, 1, 0,-1 };

typedef union {
    struct {
        unsigned x : WS;
        unsigned y : HS;
        unsigned d : 2;
    };
    unsigned v;
} state_t;

int visit[1 << (WS + HS + 2)];

state_t open[1 << 12];
int openrd, openwr;

void push(state_t s)
{
    open[openwr++] = s;
    openwr &= COUNT(open) - 1;
    assert(openrd != openwr);
}

state_t pop()
{
    assert(openrd != openwr);
    state_t s = open[openrd++];
    openrd &= COUNT(open) - 1;
    return s;
}

int bfs(state_t start, state_t end, int* count)
{
    for (int i = 0; i < COUNT(visit); i++) {
        visit[i] = INT_MAX;
    }
    int shortest = INT_MAX;
    visit[start.v] = 0;
    push(start);

    while (openwr != openrd) {
        state_t s = pop();
        int dist = visit[s.v];
        //A(s.x, s.y) = 'x';
        if (s.x == end.x && s.y == end.y) {
            if (dist < shortest) shortest = dist;
        }

        for (int dd = -1; dd <= 1; dd++) {
            state_t t = s;
            int ndist = dist;
            if (dd == 0) {
                t.x += dx[s.d];
                t.y += dy[s.d];
                if (A(t.x, t.y) == '#') continue;
                ndist += 1;
            } else {
                t.d = (s.d + dd + 4) % 4;
                ndist += 1000;
            }
            if (ndist < visit[t.v]) {
                visit[t.v] = ndist;
                push(t);
            }
        }
    }

    *count = 0;
    for (int d = 0; d < 4; d++) {
        state_t t = end;
        t.d = d;
        if (visit[t.v] == shortest) {
            push(t);
        }
    }
    while (openwr != openrd) {
        state_t t = pop();
        int dist = visit[t.v];
        if (A(t.x, t.y) != 'O') {
            A(t.x, t.y) = 'O';
            (*count)++;
        }

        for (int dd = -1; dd <= 1; dd++) {
            state_t s = t;
            int ndist = dist;
            if (dd == 0) {
                s.x -= dx[t.d];
                s.y -= dy[t.d];
                ndist -= 1;
            } else {
                s.d = (t.d + dd + 4) % 4;
                ndist -= 1000;
            }
            if (visit[s.v] == ndist) {
                push(s);
            }
        }
    }

    return shortest;
}

int main()
{
    FILE* file = fopen("day16.txt", "r");
    int sum = 0, sum2 = 0;

    while (1 == fscanf(file, "%s", &A(0, n))) n++;

    int sx, sy, ex, ey;
    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            if (A(x, y) == 'S') {
                A(x, y) = '.';
                sx = x, sy = y;
            } else if (A(x, y) == 'E') {
                A(x, y) = '.';
                ex = x, ey = y;
            }
        }
    }
    //printf("n %d sx %d sy %d ex %d ey %d\n", n, sx, sy, ex, ey);

    sum = bfs((state_t){{ sx, sy, 0 }}, (state_t){{ ex, ey, 0 }}, &sum2);

    for (int y = 0; y < n; y++) {
        for (int x = 0; x < n; x++) {
            //printf("%c", A(x, y));
        }
        //printf("\n");
    }

    printf("%d %d\n", sum, sum2);
}

File: day16.lua
#!/usr/bin/env lua

local a = {}

local dx = { 1, 0,-1, 0 }
local dy = { 0, 1, 0,-1 }

local function key(self)
    return self.x | self.y << 8 | self.d << 16
end

local function state(x, y, d)
    return {
        x = x,
        y = y,
        d = d,
        v = key,
    }
end

local open = {}
local openrd, openwr = 1, 1

local function push(s)
    open[openwr] = s
    openwr = openwr + 1
end

local function pop()
    assert(openrd ~= openwr)
    local s = open[openrd]
    open[openrd] = nil
    openrd = openrd + 1
    return s
end

local function bfs(start, ends)
    local visit = setmetatable({}, { __index = function() return math.maxinteger end })
    local shortest = math.maxinteger
    visit[start:v()] = 0
    push(start)

    while openrd ~= openwr do
        local s = pop()
        local dist = visit[s:v()]
        --a[s.y][s.x] = 'x'
        if s.x == ends.x and s.y == ends.y then
            if dist < shortest then shortest = dist end
        end

        for dd = -1, 1 do
            local t = state(s.x, s.y, s.d)
            local ndist = dist
            if dd == 0 then
                t.x = s.x + dx[s.d]
                t.y = s.y + dy[s.d]
                ndist = ndist + 1
            else
                t.d = ((s.d - 1 + dd + 4) % 4) + 1
                ndist = ndist + 1000
            end
            if a[t.y][t.x] ~= '#' and ndist < visit[t:v()] then
                visit[t:v()] = ndist
                push(t)
            end
        end
    end

    local count = 0
    for d = 1, 4 do
        local t = state(ends.x, ends.y, d)
        if visit[t:v()] == shortest then
            push(t)
        end
    end
    while openrd ~= openwr do
        local t = pop()
        local dist = visit[t:v()]
        if a[t.y][t.x] ~= 'O' then
            a[t.y][t.x] = 'O'
            count = count + 1
        end

        for dd = -1, 1 do
            local s = state(t.x, t.y, t.d)
            local ndist = dist
            if dd == 0 then
                s.x = t.x - dx[t.d]
                s.y = t.y - dy[t.d]
                ndist = ndist - 1
            else
                s.d = ((t.d - 1 + dd + 4) % 4) + 1
                ndist = ndist - 1000
            end
            if visit[s:v()] == ndist then
                push(s)
            end
        end
    end

    return shortest, count
end

local file = io.open("day16.txt")

for line in file:lines() do
    local row = {}
    for i = 1, #line do row[i] = line:sub(i, i) end
    table.insert(a, row)
end

local sx, sy, ex, ey
for y = 1, #a do
    for x = 1, #a[y] do
        if a[y][x] == 'S' then
            a[y][x] = '.'
            sx, sy = x, y
        elseif a[y][x] == 'E' then
            a[y][x] = '.'
            ex, ey = x, y
        end
    end
end
--print("n", #a, "sx", sx, "sy", sy, "ex", ex, "ey", ey)

local sum, sum2 = bfs(state(sx, sy, 1), state(ex, ey, 1))

for y = 1, #a do
    for x = 1, #a[y] do
        --io.write(a[y][x])
    end
    --print()
end

print(sum, sum2)
